File: src/main/webapp/js/inputprompt.js
// Observe a text input field, inserting an "input prompt" when
// it is empty and does not have focus. Written in Prototype style because
// Tapestry includes the Prototype library (http://www.prototypejs.org/).
//
// Adapted from http://jumpstart.doublenegative.com.au/jumpstart/examples/javascript/mixin

InputPrompt = Class.create( {

	initialize : function(textboxId, promptText, promptColor) {
		this.textbox = $(textboxId);
		this.promptText = promptText;
		this.promptColor = promptColor;
		this.normalColor = this.textbox.getStyle('color');

		Event.observe(this.textbox, 'focus', this.doClearPrompt.bindAsEventListener(this));
		Event.observe(this.textbox, 'blur', this.doCheckPrompt.bindAsEventListener(this));
		Event.observe(this.textbox, 'change', this.doCheckPrompt.bindAsEventListener(this));
		Event.observe(this.textbox.form, 'submit', this.doClearPrompt.bindAsEventListener(this));
		
		this.doCheckPrompt();
	},
	
	doClearPrompt : function(e) {
		if (this.textbox.value == this.promptText) {
			this.textbox.value = "";
		}
		this.textbox.setStyle({color: this.normalColor});
	},

	doCheckPrompt : function(e) {

		// If field is empty, put the promptText in it and set its color
		if (this.textbox.value.length == 0) {
			this.textbox.value = this.promptText;
			this.textbox.setStyle({color: this.promptColor});
		}
		// Else if field contains promptText, set its color to promptColor
		else if (this.textbox.value == this.promptText) {
			this.textbox.setStyle({color: this.promptColor});
		}
		// Else, set the field's color to its normal color
		else {
			this.textbox.setStyle({color: this.normalColor});
		}
	}
} )

// Extend Tapestry.Initializer with static method that instantiates an InputPrompt

Tapestry.Initializer.InputPrompt = function(spec) {
	new InputPrompt(spec.textboxId, spec.promptText, spec.promptColor);
}

File: src/main/java/org/apache/tapestry/finder/mixins/InputPrompt.java
/**
 * This mixin uses JavaScript to observe a field, inserting prompt text
 * into it whenever it is empty and does not have focus.
 * 
 * This is from http://jumpstart.doublenegative.com.au/jumpstart/examples/javascript/mixin
 */

package org.apache.tapestry.finder.mixins;

import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.Import;
import org.apache.tapestry5.annotations.InjectContainer;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.ioc.annotations.Inject;
import org.apache.tapestry5.json.JSONObject;
import org.apache.tapestry5.services.javascript.JavaScriptSupport;

@Import(library = "${context:/js/inputprompt.js")
public class InputPrompt
{
	// Required parameter: the text to insert as a prompt
	@Parameter(required = true, defaultPrefix = BindingConstants.LITERAL)
	private String promptText;

	// Optional parameter: color of the prompt
	@Parameter(defaultPrefix = BindingConstants.LITERAL)
	private String promptColor;

	@Inject
	private JavaScriptSupport javaScriptSupport;

	@InjectContainer
	private ClientElement clientElement;

	public void setupRender()
	{
		promptColor = "#808080";	// default text color for prompt
	}

	/**
	 * Tell the Tapestry.Initializer to set up the InputPrompt (after the
	 * DOM has been fully loaded). 
	 */
	@AfterRender
	public void addInitializer()
	{
		JSONObject spec = new JSONObject();
		spec.put("textboxId", clientElement.getClientId());
		spec.put("promptText", promptText);
		spec.put("promptColor", promptColor);
		javaScriptSupport.addInitializerCall("inputPrompt", spec);
	}

}